### DISCUSSION TITLE
Can't Re open workflow ,after Remote Hop-Server restart

### DISCUSSION THREAD

**Discussion by KennyChuang2000** [apache/hop#4596] [Q&A] [answered]
Hello everyone,

I use apache/hop image to start a  container to be my remote Hop Server, then I execute  a workflow  in  Hop-gui with remote Hop Server .
It works! I can manage my workflow on  remote Hop Server, I am trying to findout how to keep old workflow on  remote Hop Server after the container close or restart . but i can't find any .hwf or .hpl file in my container.

1.I want to know where the .hwf file will be uploaded to the container after I execute the workflow on a remote server?

2.Can I change the .hwf storage location so that the Hop server can re-execute the previously executed .hwf files after container restart?

Warm regards,
Kenny

**Comment by bamaer** [comment]
Hop Server is stateless. It won't remember any of your previous workflows, pipelines or their executions after a restart.

**Reply by KennyChuang2000** [reply]
Thank you for reply ! 
But i still not understand that if .hwf file didn't  exists in  Hop-Sever's container ,how can container execute workflow?(container didn't restart)

**Reply by bamaer** [reply, marked as answer]
not sure I understand your question correctly. what I understand from your initial question is that you started your workflow from Hop Gui. That will send the workflow (and linked resources if that option is checked) to the server before it is executed. You would need to repeat that process through Hop Gui or the REST if you want to repeat that process after a restart.

**Reply by KennyChuang2000** [reply]
Thank you for your suggest, After reading your suggestions, I am trying to use Rest Api  to  upload my workflow 
```
curl -X POST "http://localhost:8081/hop/addWorkflow/xml=Y" -H "Content-Type: application/xml" --data-binary "@D:\\Apach Hop\\apache-hop-client-2.10.0\\hop\\testqWorkflow.xml" -u cluster:cluster
```
I jsut  change my testqWorkflow.hwf  File extension to testqWorkflow.xml ，and error like this 
```
org.apache.hop.core.exception.HopException:
Error reading metadata from JSON

Error find metadata class for key null

The metadata plugin for key null could not be found in the plugin registry
``` 
it means I have attchch "matadata" folder in this api request?

**Reply by bamaer** [reply]
have a look at the [registerPipeline](https://hop.apache.org//manual/latest/hop-server/rest-api.html#_register_pipeline) example, it's similar for registerWorkflow.  you could also consider running short-lived containers if you don't have a use case that has a hard requirement for a 24/7 server

**Reply by KennyChuang2000** [reply]
I manually changed the .hpl file extension to .xml and edited the .xml to add
</pipeline_execution_configuration> &</metastore_json> tag in the example, the result is successfully addPipeline to hop-server!

Does hop-gui currently have the function of automatically exporting XML that conforms to the REST API format?

**Reply by bamaer** [reply]
yes, but that export happens in the background when publishing to the server, the XML isn't exposed.
